src/App/Controls/PasswordStrengthProgressBar/PasswordStrengthProgressBar.xaml.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Xamarin.Forms;
+
+namespace Bit.App.Controls
+{
+    public partial class PasswordStrengthProgressBar : StackLayout
+    {
+        private readonly PasswordStrengthViewModel _viewModel;
+
+        public static readonly BindableProperty PasswordStrengthProperty = BindableProperty.Create(
+            nameof(PasswordStrength),
+            typeof(PasswordStrengthCategory),
+            typeof(PasswordStrengthProgressBar),
+            defaultBindingMode: BindingMode.OneWayToSource);
+
+        public static readonly BindableProperty EmailProperty = BindableProperty.Create(
+            nameof(Email),
+            typeof(string),
+            typeof(PasswordStrengthProgressBar),
+            string.Empty);
+
+        public static readonly BindableProperty PasswordProperty = BindableProperty.Create(
+            nameof(Password),
+            typeof(string),
+            typeof(PasswordStrengthProgressBar),
+            string.Empty);
+
+        public PasswordStrengthCategory PasswordStrength
+        {
+            get { return (PasswordStrengthCategory)GetValue(PasswordStrengthProperty); }
+            set { SetValue(PasswordStrengthProperty, value); }
+        }
+
+        public string Email
+        {
+            get { return (string)GetValue(EmailProperty); }
+            set { SetValue(EmailProperty, value); }
+        }
+
+        public string Password
+        {
+            get { return (string)GetValue(PasswordProperty); }
+            set { SetValue(PasswordProperty, value); }
+        }
+
+        public PasswordStrengthProgressBar()
+        {
+            InitializeComponent();
+            _viewModel = new PasswordStrengthViewModel();
+            _passwordBar.BindingContext = _viewModel;
+            _passwordStatus.BindingContext = _viewModel;
+        }
+
+        protected override void OnPropertyChanged(string propertyName = null)
+        {
+            base.OnPropertyChanged(propertyName);
+            if (propertyName == PasswordProperty.PropertyName ||
+                propertyName == EmailProperty.PropertyName)
+            {
+                _viewModel.CalculateMasterPasswordStrength(Password, Email);
+                _passwordBar.ProgressTo(_viewModel.PasswordStrength, 500, Easing.SinOut);
+                PasswordStrength = _viewModel.PasswordStrengthCategory;
+            }
+        }
```

Comment:
I think you could improve the approach and end up simplifying the number of bindings required by moving away the ViewModel initialization to the host ViewModel of the host view instead of initializing everything here. Then the host view can bind the `PasswordStrengthProgressBar` binding context to this ViewModel, therefore you wouldn't have to need all these binding properties. Taking as an example the `RegisterPage`:

We'd need an interface to use IoC (I use `UserInputs` instead of `Email` because on other places that this progress bar can be used there might be other user inputs and in fact is the thing required by the service):
```c#
public interface IPasswordStrengthable
{
    string Password { get; }
    List<string> UserInputs { get; }
}
```

On `RegisterPageViewModel` (I pass itself so it can implement the interface needed for it to work using IoC):
```c#
public class RegisterPageViewModel : CaptchaProtectedViewModel, IPasswordStrengthable
{
    ...
    public RegisterPageViewModel(...)
    {
        ...
        PasswordStrengthViewModel = new PasswordStrengthViewModel(this);
    }

    public PasswordStrengthViewModel PasswordStrengthViewModel { get; }

    public string Password => MasterPassword;
    
    public List<string> UserInputs => PasswordStrengthViewModel.GetEmailUserInput(Email);
```

On `RegisterPage.xaml`:
```xaml
<controls:PasswordStrengthProgressBar
    BindingContext={Binding PasswordStrengthViewModel}
    Margin="0,0"/>
```

and finally on the `PasswordStrengthViewModel` you can use directly the reference to whatever implements the interface to calculate the strength:

```c#
IPasswordStrengthable _passwordStrengthable

public PasswordStrengthViewModel(IPasswordStrengthable passwordStrengthable)
{
    _passwordStrengthable = passwordStrengthable;
    _passwordGenerationService = ServiceContainer.Resolve<IPasswordGenerationService>();
}

public void CalculateMasterPasswordStrength()
{
    var passwordStrength = _passwordGenerationService.PasswordStrength(_passwordStrengthable.Password, _passwordStrengthable.UserInputs);
    ...
```

IMHO, this provides a better approach, more scalable, flexible that you can use anywhere just by implementing the interface.

Reply:
I think it is all done now :)